src/pose.rs

```diff
@@ -0,0 +1,249 @@
+    poses: Arc<RwLock<Vec<Isometry3<f64>>>>,
+    _subscriber: rosrust::Subscriber,
+}
+
+impl PoseArrayListener {
+    pub fn new(config: PoseListenerConfig) -> PoseArrayListener {
+        let poses = Arc::new(RwLock::new(Vec::<Isometry3<f64>>::new()));
+        let cb_poses = poses.clone();
+        let sub = rosrust::subscribe(
+            &config.topic,
+            2,
+            move |pose_array: rosrust_msg::geometry_msgs::PoseArray| {
+                let poses_iso = pose_array
+                    .poses
+                    .into_iter()
+                    .map(|p| ros_pose_to_isometry(&p))
+                    .collect();
+                *cb_poses.write().unwrap() = poses_iso;
+            },
+        )
+        .unwrap();
+
+        PoseArrayListener {
+            config: config,
+            poses: poses,
+            _subscriber: sub,
+        }
+    }
+
+    pub fn get_lines(&self) -> Vec<Line> {
+        if self.poses.read().unwrap().is_empty() {
+            return Vec::new();
+        }
+        match self.config.style.as_str() {
+            "arrow" => self
+                .poses
+                .read()
+                .unwrap()
+                .iter()
+                .map(|p| pose_to_arrow(&p, self.config.length, &self.config.color))
+                .collect::<Vec<Vec<Line>>>()
+                .into_iter()
+                .reduce(|a, b| a.into_iter().chain(b.into_iter()).collect())
+                .unwrap(),
+            "axis" => self
+                .poses
+                .read()
+                .unwrap()
+                .iter()
+                .map(|p| pose_to_axes(&p, self.config.length))
+                .collect::<Vec<Vec<Line>>>()
+                .into_iter()
+                .reduce(|a, b| a.into_iter().chain(b.into_iter()).collect())
+                .unwrap(),
+            _ => Vec::new(),
+        }
+    }
+}
+
+pub struct PathListener {
+    config: PoseListenerConfig,
+    poses: Arc<RwLock<Vec<Isometry3<f64>>>>,
+    _subscriber: rosrust::Subscriber,
+}
+
+impl PathListener {
+    pub fn new(config: PoseListenerConfig) -> PathListener {
+        let poses = Arc::new(RwLock::new(Vec::<Isometry3<f64>>::new()));
+        let cb_poses = poses.clone();
+        let sub = rosrust::subscribe(
+            &config.topic,
+            2,
+            move |path: rosrust_msg::nav_msgs::Path| {
+                let poses_iso = path
+                    .poses
+                    .into_iter()
+                    .map(|p| ros_pose_to_isometry(&p.pose))
+                    .collect();
+                *cb_poses.write().unwrap() = poses_iso;
+            },
+        )
+        .unwrap();
+
+        PathListener {
+            config: config,
+            poses: poses,
+            _subscriber: sub,
+        }
+    }
+
+    pub fn get_lines(&self) -> Vec<Line> {
+        if self.poses.read().unwrap().is_empty() {
+            return Vec::new();
+        }
+        match self.config.style.as_str() {
```

Comment:
the styling is repetitive code

Reply:
I can wrap the super long series of methods if you want (having a poses_to_lines or so), but the match depends on the message (well only path has lines). How more would you suggest to reduce the code?